subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c

```diff
@@ -0,0 +1,611 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_img_grp
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h>
+
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+
+#define MCUMGR_UPLOAD_INIT_HEADER_BUF_SIZE 128
+
+/* Pointer for active Client */
+static struct img_mgmt_client *active_client;
+/* Image State read or set response pointer */
+static struct mcumgr_image_state *image_info;
+/* Image upload response pointer */
+static struct mcumgr_image_upload *image_upload_buf;
+
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static const char smp_images_str[] = "images";
+#define IMAGES_STR_LEN (sizeof(smp_images_str) - 1)
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size)
+{
+	for (int i = 0; i < map_size; i++) {
+		/* Clear found status */
+		map[i].found = false;
+	}
+}
+
+static int image_state_res_fn(struct net_buf *nb, void *user_data)
+{
+	zcbor_state_t zsd[CONFIG_MCUMGR_SMP_CBOR_MAX_DECODING_LEVELS + 2];
```

Comment:
+2 is for backup state purpose.  It was already used at server side so I follow same path.

Reply:
Ah, I see. I can only find [this single line](https://github.com/zephyrproject-rtos/zephyr/blob/9d6ced38226146071369ebc4bca9e78b8aa4786f/include/zephyr/mgmt/mcumgr/smp/smp.h#L51) which has the `+ 2` and also a comment. I think it will be difficult to figure out that the client code took the logic from there, thus difficult to figure out where the `+ 2` comes from.